lib/libutee/arch/arm/ftrace/ftrace.c

```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2019, Linaro Limited
+ */
+
+#include <string.h>
+#include <user_ta_header.h>
+#include <utee_syscalls.h>
+#include "ftrace.h"
+
+#define FTRACE_RETFUNC_DEPTH		50
+
+static unsigned long ftrace_ret_stack[FTRACE_RETFUNC_DEPTH];
+static unsigned int ftrace_ret_idx;
+static const char hex_str[] = "0123456789abcdef";
+
+static void __noprof ftrace_memmove(struct ftrace_buf *fbuf, size_t size)
+{
+	char *dst = (char *)fbuf + fbuf->buf_off;
+	const char *src = (char *)fbuf + fbuf->buf_off + size;
+	size_t len = 0;
+
+	fbuf->curr_size -= size;
+	len = fbuf->curr_size;
+
+	while (len--)
+		*dst++ = *src++;
+}
+
+static size_t __noprof to_func_enter_fmt(char *buf, unsigned long pc)
+{
+	char *str = buf;
+	uint32_t addr_size = 2 * sizeof(unsigned long);
+	uint32_t i = 0;
+
+	for (i = 0; i < ftrace_ret_idx; i++)
+		*str++ = ' ';
+
+	*str++ = '0';
+	*str++ = 'x';
+
+	for (i = 0; i < addr_size; i++)
+		*str++ = hex_str[(pc >> 4 * (addr_size - i - 1)) & 0xf];
+
+	*str++ = '(';
+	*str++ = ')';
+	*str++ = ' ';
+	*str++ = '{';
+	*str++ = '\n';
+	*str = '\0';
+
+	return str - buf;
+}
+
+void __noprof ftrace_enter(unsigned long pc, unsigned long *lr)
+{
+	struct ftrace_buf *fbuf = NULL;
+	size_t dump_size = 0;
+
+	fbuf = &__ftrace_buf_start;
+
+	if (!fbuf->buf_off || !fbuf->max_size)
+		return;
+
+	dump_size = ftrace_ret_idx + (2 * sizeof(unsigned long)) + 8;
+
+	if ((fbuf->curr_size + dump_size) > fbuf->max_size)
+		ftrace_memmove(fbuf, dump_size);
```

Comment:
The problem is, doing this does not guarantee the structure of the dump file, so it can happen that `symbolize.py` fails to decode the dump. In one of my tests I got this:
```
 region  0: va 0x0000000040000000 pa 0x000000000e100000 size 0x002000 flags ---R-X
 region  1: va 0x0000000040002000 pa 0x000000000e191000 size 0x001000 flags ---RW-
 region  2: va 0x0000000040004000 pa 0x000000000e455000 size 0x009000 flags rw-RW-
 region  3: va 0x000000004000d000 pa 0x000000000e362000 size 0x05f000 flags r-x--- [0]
 region  4: va 0x000000004006c000 pa 0x000000000e3c1000 size 0x094000 flags rw---- [0]
 [0] cb3e5ba0-adf1-11e0-998b-0002a5d5c51b @ 0x000000004000d000
Function graph for TA: cb3e5ba0-adf1-11e0-998b-0002a5d5c51b @ 0x4000d000
58c8();
  }
  0x0000000040012398() {
   0x000000004000f3f8() {
    0x000000004001b8b4() {
     0x0000000040019200() {
      0x0000000040055f94() {
[...]
```

This cannot be decoded, but if I simply remove the line after "Function graph", then it works.
I don't know if this should be addressed here or in `symbolize.py`, perhaps in the Python script is easier and more robust.

Reply:
Hmm, I see. Yes will make `symbolize.py` robust to handle this case.

Reply:
Just a thought, do we have to do all dump analysis using `symbolize.py`? It's maybe easier and more clean to keep things in separate scripts?

Reply:
Currently for aborts and ftrace dumps, we utilize most of common framework to symbolize function addresses. So are you suggesting to create separate scripts (with duplicate code) or create separate wrapper scripts to reuse common framework?

BTW, I think this change should be taken as separate patch from this PR.

Reply:
> So are you suggesting to create separate scripts (with duplicate code) or create separate wrapper scripts to reuse common framework?

Not really, just raising that as a thought.

> BTW, I think this change should be taken as separate patch from this PR.

Totally agree on that!